src/matrix.c

```diff
@@ -0,0 +1,232 @@
+#include "data.table.h"
+
+/* To handle different matrix types (e.g. numeric, character, etc)
+ * there is a single dispatch function asmatrix at the bottom of the 
+ * file that detects the appropriate R atomic type then calls the
+ * corresponding asmatrix_<type> function.
+ */
+
+SEXP asmatrix_logical(SEXP dt) {
+  // Determine the number of rows and columns in the data.table
+  R_len_t p = length(dt);
+  R_len_t n = length(VECTOR_ELT(dt, 0));
+  
+  // Create output matrix, allocate memory, and create a pointer to it.
+  SEXP mat = PROTECT(allocMatrix(LGLSXP, n, p));
+  int *pmat; 
+  pmat = LOGICAL(mat);
+  
+  /* Create an array of pointers for the individual columns in DT 
+   * so that we avoid creating these and calling the helper functions
+   * in the parallel OMP threads
+   */
+  int **pcol;
+  pcol = (int **) R_alloc(p, sizeof(int *));
+  for (R_len_t jj = 0; jj < p; jj++) {
+    pcol[jj] = LOGICAL(VECTOR_ELT(dt, jj));
+  }
+
+  // Multithreaded for loop to copy each column into the matrix memory location
+  #pragma omp parallel for num_threads(getDTthreads())
+  for (R_len_t jj = 0; jj < p; jj++) {
+    memcpy(pmat + (jj * n), pcol[jj], sizeof(int)*n);
```

Comment:
Another option as an alternative here is perhaps `Rf_lazy_duplicate()`, we wouldn't be able to use omp presumably, but it might save a lot of memory?